nucypher-core/src/dkg.rs

```diff
@@ -78,6 +94,129 @@ impl<'a> ProtocolObjectInner<'a> for ThresholdDecryptionRequest {
 
 impl<'a> ProtocolObject<'a> for ThresholdDecryptionRequest {}
 
+/// A request for an Ursula to derive a decryption share that specifies the key to encrypt Ursula's response.
+#[derive(PartialEq, Debug, Clone, Serialize, Deserialize)]
+pub struct E2EEThresholdDecryptionRequest {
+    /// The decryption request.
+    pub decryption_request: ThresholdDecryptionRequest,
+    /// The key to encrypt the corresponding decryption response.
+    pub response_encrypting_key: PublicKey,
+}
+
+impl E2EEThresholdDecryptionRequest {
+    /// Create E2E decryption request.
+    pub fn new(
+        decryption_request: &ThresholdDecryptionRequest,
+        response_encrypting_key: &PublicKey,
+    ) -> Self {
+        Self {
+            decryption_request: decryption_request.clone(),
+            response_encrypting_key: *response_encrypting_key,
+        }
+    }
+}
+
+impl<'a> ProtocolObjectInner<'a> for E2EEThresholdDecryptionRequest {
+    fn version() -> (u16, u16) {
+        (1, 0)
+    }
+
+    fn brand() -> [u8; 4] {
+        *b"E2eR"
+    }
+
+    fn unversioned_to_bytes(&self) -> Box<[u8]> {
+        messagepack_serialize(&self)
+    }
+
+    fn unversioned_from_bytes(minor_version: u16, bytes: &[u8]) -> Option<Result<Self, String>> {
+        if minor_version == 0 {
+            Some(messagepack_deserialize(bytes))
+        } else {
+            None
+        }
+    }
+}
+
+impl<'a> ProtocolObject<'a> for E2EEThresholdDecryptionRequest {}
+
+/// An encrypted request for an Ursula to derive a decryption share.
+#[derive(PartialEq, Debug, Clone, Serialize, Deserialize)]
+pub struct EncryptedThresholdDecryptionRequest {
+    /// ID of the ritual
+    pub ritual_id: u16,
+    /// TODO Umbral for now - but change
+    capsule: Capsule,
+    #[serde(with = "serde_bytes::as_base64")]
+    ciphertext: Box<[u8]>,
+}
+
+impl EncryptedThresholdDecryptionRequest {
+    fn new(
+        request: &ThresholdDecryptionRequest,
+        request_encrypting_key: &PublicKey,
+        response_encrypting_key: &PublicKey,
+    ) -> Self {
+        let e2e_decryption_request =
+            E2EEThresholdDecryptionRequest::new(request, response_encrypting_key);
+        // TODO: using Umbral for encryption to avoid introducing more crypto primitives.
+        let (capsule, ciphertext) =
+            match encrypt(request_encrypting_key, &e2e_decryption_request.to_bytes()) {
+                Ok(result) => result,
+                Err(err) => match err {
+                    // For now this is the only error that can happen during encryption,
+                    // and there's really no point in propagating it.
+                    EncryptionError::PlaintextTooLarge => {
+                        panic!("encryption failed - out of memory?")
+                    }
+                },
+            };
```

Comment:
If we want to convert the error or handle it in some specific way we can call `.map_error()` on `Result` type. But I think in this case it should be enough to unwrap the error and panic using `.expect()`
```suggestion
        let (capsule, ciphertext) = encrypt(request_encrypting_key, &e2e_decryption_request.to_bytes())
          .expect("encryption failed - out of memory?");
```

Reply:
✔️